src/main/java/org/sonarlint/intellij/issue/IssueStore.java

```diff
@@ -101,55 +108,92 @@ public void clean(VirtualFile file) {
-    for (Map.Entry<VirtualFile, Collection<IssuePointer>> e : map.entrySet()) {
+  public void store(Map<VirtualFile, Collection<LocalIssuePointer>> map) {
+    for (Map.Entry<VirtualFile, Collection<LocalIssuePointer>> e : map.entrySet()) {
       store(e.getKey(), e.getValue());
     }
     messageBus.syncPublisher(IssueStoreListener.SONARLINT_ISSUE_STORE_TOPIC).filesChanged(map);
   }
 
   private void cleanInvalid(VirtualFile file) {
-    Collection<IssuePointer> issues = storePerFile.get(file);
+    Collection<LocalIssuePointer> issues = storePerFile.get(file);
     if (issues == null) {
       return;
     }
 
-    Iterator<IssuePointer> it = issues.iterator();
+    Iterator<LocalIssuePointer> it = issues.iterator();
     while (it.hasNext()) {
       if (!it.next().isValid()) {
         it.remove();
       }
     }
   }
 
-  void store(VirtualFile file, final Collection<IssuePointer> rawIssues) {
+  void store(VirtualFile file, final Collection<LocalIssuePointer> rawIssues) {
     boolean firstAnalysis = !storePerFile.containsKey(file);
 
     // clean before issue tracking
     cleanInvalid(file);
 
     // this will also delete all existing issues in the file
-    if(firstAnalysis) {
+    if (firstAnalysis) {
       // don't set creation date, as we don't know when the issue was actually created (SLI-86)
       storePerFile.put(file, rawIssues);
     } else {
-      final Collection<IssuePointer> previousIssues = getForFile(file);
-      Collection<IssuePointer> trackedIssues = new ArrayList<>();
-
-      Input<IssuePointer> baseInput = () -> previousIssues;
-      Input<IssuePointer> rawInput = () -> rawIssues;
-      Tracking<IssuePointer, IssuePointer> tracking = new Tracker<IssuePointer, IssuePointer>().track(rawInput, baseInput);
-      for (Map.Entry<IssuePointer, IssuePointer> entry : tracking.getMatchedRaws().entrySet()) {
-        IssuePointer rawMatched = entry.getKey();
-        IssuePointer previousMatched = entry.getValue();
-        rawMatched.setCreationDate(previousMatched.creationDate());
-        trackedIssues.add(rawMatched);
-      }
-      for (IssuePointer newIssue : tracking.getUnmatchedRaws()) {
-        newIssue.setCreationDate(System.currentTimeMillis());
-        trackedIssues.add(newIssue);
+      matchingInProgress.lock();
+      Collection<LocalIssuePointer> previousIssues = getForFile(file);
+      Collection<LocalIssuePointer> trackedIssues = new ArrayList<>();
+
+      Input<LocalIssuePointer> baseInput = () -> previousIssues;
+      Input<LocalIssuePointer> rawInput = () -> rawIssues;
+      updateTrackedIssues(file, trackedIssues, baseInput, rawInput);
+      matchingInProgress.unlock();
+    }
+  }
+
+  public void storeServerIssues(VirtualFile file, final Collection<IssuePointer> serverIssues) {
+    if (!storePerFile.containsKey(file)) {
+      // server issue gone, or file was renamed locally
+      // TODO add support to track renamed files (or explain why not)
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)